lib/tapioca/executor.rb

```diff
@@ -25,24 +28,62 @@ def initialize(queue, number_of_workers: nil)
       @items_per_worker = T.let((queue.length.to_f / @number_of_workers).ceil, Integer)
     end
 
-    sig { params(block: T.proc.params(item: T.untyped).void).void }
+    sig do
+      type_parameters(:T).params(
+        block: T.proc.params(item: T.untyped).returns(T.type_parameter(:T))
+      ).returns(T.nilable(T::Array[T.type_parameter(:T)]))
+    end
     def run_in_parallel(&block)
       # If we only have one worker selected, it's not worth forking, just run sequentially
-      if @number_of_workers == 1
-        block.call(@queue.shift) until @queue.empty?
-        return
-      end
+      return @queue.map { |item| block.call(item) }.compact if @number_of_workers == 1
+
+      # Create an IO pipe to communicate the return value of the parallelized block back from the workers to the main
+      # process. It's important to only create pipes if we're running with more than one worker, otherwise the tests
+      # fail with a "too many open files" error.
+      read, write = IO.pipe
 
       # If we have more than one worker, fork the pool by shifting the expected number of items per worker from the
       # queue
       workers = (0...@number_of_workers).map do
         items = @queue.shift(@items_per_worker)
 
-        fork { block.call(items.shift) until items.empty? }
+        fork do
+          read.close
+          result = items.map { |item| block.call(item) }.compact
+
+          # We mapped the result of invoking the parallelized block into an array. In order to return the array from the
+          # worker back to the main process, we encode it in Base64, append a separator in the beginning and write it to
+          # the pipe. The separator helps us split the results that are coming from the multiple workers. It looks
+          # something like this:
+          # -absbasd13231-asbasd123123
```

Comment:
Since all workers write to the same pipe, isn't there a slight chance that two workers write at the same time and mangle the output?

```
worker1 writes"-1111"
worker2 writes "-2222"

output: "-11-222211"
```

Reply:
I _think_ it depends on the size of what you write and if it goes over the buffer size (which I _think_ is system dependent)?

Reply:
AFAIK, the default behaviour of Ruby IO objects is to not buffer on the Ruby side and write straight to the underlying IO: https://ruby-doc.org/core-3.0.2/IO.html#sync-method

As for the underlying pipe implementation, I think what you said is true:
> Write requests of {PIPE_BUF} bytes or less shall not be interleaved with data from other processes doing writes on the same pipe. Writes of greater than {PIPE_BUF} bytes may have data interleaved, on arbitrary boundaries, with writes by other processes, whether or not the O_NONBLOCK flag of the file status flags is set.

_From: https://pubs.opengroup.org/onlinepubs/9699919799/functions/write.html_

So this really depends on the definition for `PIPE_BUF` which seems to be fairly generous for the platforms we are concerned with: https://arto.s3.amazonaws.com/notes/posix#pipe-buf

Reply:
Alternatively, we could create a pipe-per-fork and read from them sequentially in the parent process. That would be resilient to this kind of behaviour and we won't have to worry about splitting the read data, etc either.

Reply:
I have switched the implementation to use multiple pipes, I think that's nicer and we can avoid the separator. It also makes sure the return is no longer nilable, which simplifies our types a bit.